groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3548 @@
+                                std::initializer_list<INIT_LIST_TYPE> il,
+                                ARGS&&...                             args)
+: d_type(INDEX)
+, d_union(bsl::in_place_index_t<INDEX>(),
+          AllocBase::mechanism(),
+          il,
+          std::forward<ARGS>(args)...)
+{
+}
+
+template <class HEAD, class... TAIL>
+template <class TYPE, class... ARGS>
+inline
+variant<HEAD, TAIL...>::variant(bsl::allocator_arg_t,
+                                allocator_type allocator,
+                                bsl::in_place_type_t<TYPE>,
+                                ARGS&&... args)
+: AllocBase(allocator)
+, d_type(BSLSTL_VARIANT_INDEX_OF(TYPE, variant))
+, d_union(bsl::in_place_index_t<BSLSTL_VARIANT_INDEX_OF(TYPE, variant)>(),
+          AllocBase::mechanism(),
+          std::forward<ARGS>(args)...)
+{
+}
+
+template <class HEAD, class... TAIL>
+template <class TYPE, class INIT_LIST_TYPE, class... ARGS>
+inline
+variant<HEAD, TAIL...>::variant(bsl::allocator_arg_t,
+                                allocator_type allocator,
+                                bsl::in_place_type_t<TYPE>,
+                                std::initializer_list<INIT_LIST_TYPE> il,
+                                ARGS&&...                             args)
+: AllocBase(allocator)
+, d_type(BSLSTL_VARIANT_INDEX_OF(TYPE, variant))
+, d_union(bsl::in_place_index_t<BSLSTL_VARIANT_INDEX_OF(TYPE, variant)>(),
+          AllocBase::mechanism(),
+          il,
+          std::forward<ARGS>(args)...)
+{
+}
+
+template <class HEAD, class... TAIL>
+template <size_t INDEX, class... ARGS>
+inline
+variant<HEAD, TAIL...>::variant(bsl::allocator_arg_t,
+                                allocator_type allocator,
+                                bsl::in_place_index_t<INDEX>,
+                                ARGS&&... args)
+: AllocBase(allocator)
+, d_type(INDEX)
+, d_union(bsl::in_place_index_t<INDEX>(),
+          AllocBase::mechanism(),
+          std::forward<ARGS>(args)...)
+{
+}
+
+template <class HEAD, class... TAIL>
+template <size_t INDEX, class INIT_LIST_TYPE, class... ARGS>
+inline
+variant<HEAD, TAIL...>::variant(bsl::allocator_arg_t,
+                                allocator_type allocator,
+                                bsl::in_place_index_t<INDEX>,
+                                std::initializer_list<INIT_LIST_TYPE> il,
+                                ARGS&&...                             args)
+: AllocBase(allocator)
+, d_type(INDEX)
+, d_union(bsl::in_place_index_t<INDEX>(),
+          AllocBase::mechanism(),
+          il,
+          std::forward<ARGS>(args)...)
+{
+}
+#else  // BSL_VARIANT_FULL_IMPLEMENTATION
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+template <class HEAD, class... TAIL>
+inline
+variant<HEAD, TAIL...>::variant()
+: AllocBase()
+, d_type(0)
+{
+    BloombergLP::bslma::ConstructionUtil::construct(
+        reinterpret_cast<HEAD *>(BSLS_UTIL_ADDRESSOF(d_union)),
+        AllocBase::mechanism());
```

Comment:
This is also still using the questionable construction pattern; maybe it should be replaced by `emplace<0>()`.

There are also other constructors below that are still using the same pattern